File: DFS/buildTree_106.java
package com.leecode.DFS;

import java.util.HashMap;

public class buildTree_106 {
    int val;
    buildTree_106 left;
    buildTree_106 right;

    buildTree_106(int x) {
        val = x;
    }

    HashMap<Integer, Integer> map = new HashMap<>();
    int[] post;

    public buildTree_106 buildTree(int[] inorder, int[] postorder) {
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        post = postorder;
        buildTree_106 node = create(0, inorder.length - 1, 0, postorder.length - 1);
        return node;

    }

    /*
        i_S中序遍历起始地址
        i_E中序遍历结束地址
        p_S后续遍历起始地址
        p_S后续遍历结束地址

    */
    public buildTree_106 create(int i_S, int i_E, int p_S, int p_E) {
        if (i_S > i_E || p_S > p_E) {
            return null;
        }
        int root_val = post[p_E];
        //获取根节点在中序遍历中的位置
        int root_index = map.get(root_val);
        buildTree_106 node = new buildTree_106(root_val);
        /*以根节点为界限分为左右子树，左子树的中序遍历结果首地址为i_S,末地址为root_index-1;
        后序遍历结果中首地址为p_S,末地址为p_S+()root_index-i_S-1)（P_S+首地址距离根节点）

        */
        node.left = create(i_S, root_index - 1, p_S, p_S + root_index - i_S - 1);
        node.right = create(root_index + 1, i_E, p_S + root_index - i_S, p_E - 1);
        return node;
    }

}

File: DFS/buildTree_106.c
//若不理解，建议根据样例画图想一下
/*算法思想：递归

设置区间长度为mid = loc - inleft，创建结点，值为postorder[postright]
左子树的后序遍历区间范围为[postleft, postleft + mid - 1]，中序遍历区间范围为[inleft, loc - 1]
右子树的后序遍历区间范围为[postleft + mid, postright - 1]，中序遍历区间范围为[loc + 1, inright]
root -> left = create(postorder,postleft, postleft + mid - 1,inorder,inleft, loc - 1);
root -> right = create(postorder,postleft + mid, postright - 1,inorder,loc + 1, inright);

*/
struct TreeNode* create(int* postorder,int postleft, int postright, int* inorder, int inleft, int inright)
{
    if(postleft > postright || inleft > inright)
        return NULL;
    struct TreeNode* root = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root -> val = postorder[postright];
    int loc;
    //中序遍历inorder寻找值为val的元素loc
    for(int i = inleft; i <= inright; i++)
        if(inorder[i] == root -> val)
        {
            loc = i;
            break;
        }
    int mid = loc - inleft;
    root -> left = create(postorder,postleft, postleft + mid - 1,inorder,inleft, loc - 1);
    root -> right = create(postorder,postleft + mid, postright - 1,inorder,loc + 1, inright);
    return root;
}
struct TreeNode* buildTree(int* inorder, int inorderSize, int* postorder, int postorderSize){
    return create(postorder,0,postorderSize - 1,inorder,0,inorderSize - 1);
}

